onpremise/onpremise.go

```diff
@@ -0,0 +1,152 @@
+
+type Engine struct {
+	logger                        logWrapper
+	dataFile                      string
+	licenceKey                    string
+	createTempDataCopy            bool
+	dataUpdateOnStartEnabled      bool
+	dataFileUrl                   string
+	dataFilePullEveryMs           int
+	isScheduledFilePullingEnabled bool
+	loggerEnabled                 bool
+	manager                       *dd.ResourceManager
+	config                        *dd.ConfigHash
+	totalFilePulls                int
+	rdySignal                     chan struct{}
+	fileSynced                    bool
+	isManagerInitialized          bool
+}
+
+// NewResultsHash creates a new ResultsHash object
+// uaCapacity is the initial capacity of the User-Agent cache
+// overridesCapacity is the initial capacity of the overrides cache
+func (p *Engine) NewResultsHash(uaCapacity uint32, overridesCapacity uint32) *dd.ResultsHash {
+	return dd.NewResultsHash(p.manager, uaCapacity, overridesCapacity)
+}
+
+// run starts the pipeline
+func (p *Engine) run() error {
+	if len(p.dataFile) > 0 {
+		err := p.initializeManager()
+		if err != nil {
+			return err
+		}
+	}
+
+	if p.isScheduledFilePullingEnabled {
+		go p.scheduleFilePulling()
+	}
+
+	return nil
+}
+
+func (p *Engine) initializeManager() error {
+	dataFile, err := dd.GetFilePath(".", []string{p.dataFile})
+	if err != nil {
+		return fmt.Errorf("failed to get file path: %w", err)
+	}
+	p.dataFile = dataFile
+
+	err = dd.InitManagerFromFile(p.manager, *p.config, "", p.dataFile)
+	if err != nil {
+		return fmt.Errorf("failed to init manager from file: %w", err)
+	}
+	p.isManagerInitialized = true
+
+	return nil
+}
+
+func (p *Engine) createDatafileIfNotExists() error {
+	_, err := os.Stat(p.dataFile)
+	if err != nil {
+		if len(p.dataFile) == 0 {
+			p.dataFile = "data.hash"
+		}
+
+		_, err = os.Create(p.dataFile)
+		if err != nil {
+			return fmt.Errorf("failed to create data file: %w", err)
+		}
+	}
+
+	return nil
+}
+
+type PipelineOptions func(cfg *Engine) error
+
+func WithDataFile(path string) PipelineOptions {
+	return func(cfg *Engine) error {
+		cfg.dataFile = path
+		return nil
+	}
+}
+
+// WithDataUpdateUrl sets the URL to pull data from and the interval in milliseconds
+func WithDataUpdateUrl(url string, everyMs int) PipelineOptions {
+	return func(cfg *Engine) error {
+		cfg.dataFileUrl = url
+		cfg.dataFilePullEveryMs = everyMs
+		cfg.isScheduledFilePullingEnabled = true
+
+		return nil
+	}
+}
+
+// ToggleLogger enables or disables the logger
+func ToggleLogger(enabled bool) PipelineOptions {
+	return func(cfg *Engine) error {
+		cfg.logger.enabled = enabled
+		return nil
+	}
+}
+
+// WithCustomLogger sets a custom logger
+func WithCustomLogger(logger LogWriter) PipelineOptions {
+	return func(cfg *Engine) error {
+		cfg.logger = logWrapper{
+			enabled: true,
+			logger:  logger,
+		}
+
+		return nil
+	}
+
+}
+
+func New(manager *dd.ResourceManager, config *dd.ConfigHash, opts ...PipelineOptions) (*Engine, error) {
```

Comment:
I think this constructor may be a bit too explicit.  Concpetually Engine should be a higher-level API wrapper over these lower level APIs.  It should be an opaque type that provided configurations - should create all needed objects internally and be capable of doing device detection with simple method calls like:
```
engine := Engine.New(config)  // the lifetime of this object should be the lifetime of the server

//somewhere later in the code
results := engine.process(evidence)
```